### PR TITLE
Decouple from annotations

### PR DESCRIPTION
**Pull request by greg0ire**
Closes #217

### CODE REVIEW

File: lib/Doctrine/Persistence/Mapping/Driver/StaticPHPDriver.php [hunk 1]
```diff
@@ -68,7 +68,8 @@ public function loadMetadataForClass($className, ClassMetadata $metadata)
     /**
      * {@inheritDoc}
      *
-     * @todo Same code exists in AnnotationDriver, should we re-use it somehow or not worry about it?
+     * @todo Same code exists in ColocatedMappingDriver, should we re-use it
+     * somehow or not worry about it?
```
Comment (greg0ire) [inline lib/Doctrine/Persistence/Mapping/Driver/StaticPHPDriver.php:72]: This is an interesting comment. The code seems to have diverged a bit though.

File: lib/Doctrine/Persistence/Mapping/Driver/ColocatedMappingDriver.php [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+use function preg_match;
+use function preg_quote;
+use function realpath;
+use function str_replace;
+use function strpos;
+
+/**
+ * The ColocatedMappingDriver reads the mapping metadata located near the code.
+ */
+abstract class ColocatedMappingDriver implements MappingDriver
```
Comment (derrabus) [inline lib/Doctrine/Persistence/Mapping/Driver/ColocatedMappingDriver.php]: I wonder if we should make this a trait instead.

`ColocatedMappingDriver` as a type does not make sense; We'll probably never see `ColocatedMappingDriver` in a return type or `instanceof` check for instance. And all this class contains is just some shared implementation.

Comment (greg0ire) [reply lib/Doctrine/Persistence/Mapping/Driver/ColocatedMappingDriver.php]: Good point, I hadn't considered that. The fact that that class is hard to name is probably a hint at the fact that there should not be a type.

**Comment by greg0ire** [comment]
ping @derrabus :pray:

File: lib/Doctrine/Persistence/Mapping/Driver/ColocatedMappingDriver.php [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+     */
+    protected $classNames;
+
+    /**
+     * @param string ...$paths One or multiple paths where mapping classes can be found.
+     */
+    public function __construct(string ...$paths)
+    {
+        $this->addPaths($paths);
+    }
```
Comment (derrabus) [inline lib/Doctrine/Persistence/Mapping/Driver/ColocatedMappingDriver.php]: I would not move the constructor into the trait. `AnnotationDriver` for instance skips it already.

File: lib/Doctrine/Persistence/Mapping/MappingException.php [hunk 4]
```diff
@@ -28,12 +28,16 @@ public static function classNotFoundInNamespaces($className, $namespaces)
     }
 
     /**
+     * @param class-string $driverClassName
+     *
      * @return self
      */
-    public static function pathRequired()
+    public static function pathRequired(string $driverClassName)
```
Comment (derrabus) [inline lib/Doctrine/Persistence/Mapping/MappingException.php]: This is a BC break. Shall we add a new named constructor (e.g. `pathRequiredForDriver`) instead?

File: tests/Doctrine/Tests/Persistence/Mapping/ColocatedMappingDriverTest.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+use PHPUnit\Framework\TestCase;
+
+class ColocatedMappingDriverTest extends TestCase
+{
+    /**
+     * @dataProvider pathProvider
+     */
+    public function testGetAllClassNames(string $path): void
+    {
+        $driver = new MyDriver($path);
```
Comment (derrabus) [inline tests/Doctrine/Tests/Persistence/Mapping/ColocatedMappingDriverTest.php]: Shall we move the constructor call into a protected method that `AnnotationDriverTest` can override?

File: lib/Doctrine/Persistence/Mapping/Driver/AnnotationDriver.php [hunk 6]
```diff
@@ -82,57 +40,10 @@ abstract class AnnotationDriver implements MappingDriver
     public function __construct($reader, $paths = null)
```
Comment (derrabus) [inline lib/Doctrine/Persistence/Mapping/Driver/AnnotationDriver.php:40]: Should the constructor trigger a runtime deprecation?

**Review by derrabus — CHANGES_REQUESTED** [review]

**Review by derrabus — APPROVED** [review]

File: docs/en/reference/index.rst [hunk 7]
```diff
@@ -174,8 +174,8 @@ That may be in files, annotations, yaml, xml, etc.
         public function isTransient($className);
     }
 
-The Doctrine Persistence project offers a few base implementations that make it easy to implement your own XML,
-Annotations or YAML drivers.
+The Doctrine Persistence project offers a few base implementations that
+make it easy to implement your own XML, Attributes or YAML drivers.
```
Comment (malarzm) [inline docs/en/reference/index.rst:178]: Attributes, or YAML

File: docs/en/reference/index.rst [hunk 8]
```diff
@@ -233,45 +233,9 @@ AnnotationDriver
 
 .. note::
 
-    This driver requires the ``doctrine/annotations`` project. You can install it with composer.
+    This driver requires the ``doctrine/annotations`` project and is
+    deprecated because of that.
```
Comment (malarzm) [inline docs/en/reference/index.rst:237]: I'd add a note about attributes superseding annotations

Comment (malarzm) [reply docs/en/reference/index.rst:237]: Also why removing the docs below? The feature is deprecated but documentation is still valuable :)

File: lib/Doctrine/Persistence/Mapping/Driver/ColocatedMappingDriver.php [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+use function preg_match;
+use function preg_quote;
+use function realpath;
+use function str_replace;
+use function strpos;
+
+/**
+ * The ColocatedMappingDriver reads the mapping metadata located near the code.
+ */
+trait ColocatedMappingDriver
```
Comment (malarzm) [inline lib/Doctrine/Persistence/Mapping/Driver/ColocatedMappingDriver.php:28]: why trait instead of an abstract class?

**Review by malarzm — COMMENTED** [review]
I'm sorry I'm late to the party, left few minor comments

File: lib/Doctrine/Persistence/Mapping/Driver/ColocatedMappingDriver.php (hunk 9 above)
Comment (derrabus) [reply lib/Doctrine/Persistence/Mapping/Driver/ColocatedMappingDriver.php:28]: There's a resolved comment about that in this PR.

Comment (malarzm) [reply lib/Doctrine/Persistence/Mapping/Driver/ColocatedMappingDriver.php:28]: Ah thanks, haven't seen that :)

File: docs/en/reference/index.rst (hunk 8 above)
Comment (greg0ire) [reply docs/en/reference/index.rst:237]: I think it feels weird to deprecate a feature yet help people build on it. I thought I would make it harder for people to shoot themselves in the foot.

**Pull request merged**